kvcoord: key range cache entries by regular keys, not meta keys

andreimatei opened the pull request:
Before this patch, entries in the range cache were keyed by
RangeMetaKey(desc.EndKey). There were no good reasons why either
the RangeMetaKey() transformation was used, or why the EndKey was used
instead of the StartKey. Presumably this has all been done in order to
match how descriptors are stored in the meta ranges, but this was
misguided since it's confusing.

This patch makes cache entries simply be keyed on desc.Start. This fixes
a problem with querying the cache with RKeyMin.Next(). Before this
patch, this query was broken since it was failing to find the cached
Meta1 range: the range was [RKeyMin, Meta2Start), and so it was keyed at
RangeMetaKey(Meta2Start) = Meta1/Meta2Start. The cache lookup was using
RangeMetaKey(RKeyMin.Next), which results in /Meta2/<something> (which
is too high) because of arguably a bug in the implementation of
RangeMetaKey. Fixing that bug proves to not be quite straigh-forward
(#52786). This patch fixes the problem by not needing to apply neither
.Next() nor RangeMetaKey() when performing cache lookups.

This patch fixes the issues referenced below; what the restores in those
issues were experiencing was a spectacular consequence of the failure to
lookup the cache described above: evicting the Meta1 descriptor from the
cache was always failing (because the lookup was failing) which, in
turn, was causing the DistSender to believe that it was perpetually
receiving old lease information and so it would loop around contacting
the same replica over and over. There's something to improve in the
DistSender, coming separately.
While the underlying RangeMetaKey() bug is old, I believe this bug was
exposed more by the recent range cache changes. It used to be that we
would generally evict the Meta1 descriptor from the cache based on the
key that caused that descriptor to be inserted in the cache. Now we
generally evict the descriptor based on the range's start key (which is
RKeyMin, and suffers from the RKeyMin.Next() issue).

Fixes #52468
Fixes #52391
Fixes #50987

Release note: None

andreimatei committed:
kvclient: rename range cache func Locked->RLocked

Release note: None

andreimatei committed:
kvclient: plumb some ctx

Release note: None

andreimatei committed:
kvcoord: key range cache entries by regular keys, not meta keys

Before this patch, entries in the range cache were keyed by
RangeMetaKey(desc.EndKey). There were no good reasons why either
the RangeMetaKey() transformation was used, or why the EndKey was used
instead of the StartKey. Presumably this has all been done in order to
match how descriptors are stored in the meta ranges, but this was
misguided since it's confusing.

This patch makes cache entries simply be keyed on desc.Start. This fixes
a problem with querying the cache with RKeyMin.Next(). Before this
patch, this query was broken since it was failing to find the cached
Meta1 range: the range was [RKeyMin, Meta2Start), and so it was keyed at
RangeMetaKey(Meta2Start) = Meta1/Meta2Start. The cache lookup was using
RangeMetaKey(RKeyMin.Next), which results in /Meta2/<something> (which
is too high) because of arguably a bug in the implementation of
RangeMetaKey. Fixing that bug proves to not be quite straigh-forward
(#52786). This patch fixes the problem by not needing to apply neither
.Next() nor RangeMetaKey() when performing cache lookups.

This patch fixes the issues referenced below; what the restores in those
issues were experiencing was a spectacular consequence of the failure to
lookup the cache described above: evicting the Meta1 descriptor from the
cache was always failing (because the lookup was failing) which, in
turn, was causing the DistSender to believe that it was perpetually
receiving old lease information and so it would loop around contacting
the same replica over and over. There's something to improve in the
DistSender, coming separately.
While the underlying RangeMetaKey() bug is old, I believe this bug was
exposed more by the recent range cache changes. It used to be that we
would generally evict the Meta1 descriptor from the cache based on the
key that caused that descriptor to be inserted in the cache. Now we
generally evict the descriptor based on the range's start key (which is
RKeyMin, and suffers from the RKeyMin.Next() issue).

Fixes #52468
Fixes #52391
Fixes #50987

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52817)
<!-- Reviewable:end -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> although now that I see how small this is, I'm less convinced that it's worth it in its entirety. Avoiding the `RangeMetaKey` translation when inserting and retrieving is a nice win, let's definitely keep that. Storing descriptors by their StartKey instead of their EndKey is a little less of a clear win. Conceptually it's nicer, but it actually makes the code more complex, not less complex like I was expecting, because we can't just `.Next()` to get around inclusive vs. exclusive discrepancies.

So all things considered, I think my vote would be to store descriptors by `EndKey`, instead of `RangeMetaKey(EndKey)` (current behavior) or `StartKey` (this behavior). I'm curious about what you think.

> Reviewed 2 of 2 files at r1, 6 of 6 files at r2, 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52817#-:-MEyKe957J4EMDn50AGv:b1ne7wn)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/kv/kvclient/kvcoord/range_cache.go, line 558 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/52817#-MEx_B96-bmsSbAMxkB9:-MEx_B96-bmsSbAMxkBA:b-5n8rn8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a825ae7ee001f57770a8089b922a4680903918b1/pkg/kv/kvclient/kvcoord/range_cache.go#L558)):*
> ```Go
> 	end := rangeCacheKey(span.EndKey)
> 	// Align the span's start on a range boundary.
> 	floor, _ /* rawEntry */ := rdc.getCachedRLocked(ctx, span.Key, false /* inverted */)
> ```

Instead of scanning backwards and then forwards, can we scan backwards from the EndKey and retain the single pass?
___
*[pkg/kv/kvclient/kvcoord/range_cache.go, line 908 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/52817#-MEyJfd398pohUaBgJaY:-MEyJfd47cKJbmuBIzU-:bmud76o) ([raw file](https://github.com/cockroachdb/cockroach/blob/a825ae7ee001f57770a8089b922a4680903918b1/pkg/kv/kvclient/kvcoord/range_cache.go#L908)):*
> ```
> rangeCacheKey(roachpb.RKeyMin)
> ```

Should we do the same thing we did with `maxCacheKey`?
___
*[pkg/util/cache/cache.go, line 462 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/52817#-MEyIEWoBKyHlLQeElDG:-MEyIEWoBKyHlLQeElDH:b856eki) ([raw file](https://github.com/cockroachdb/cockroach/blob/a825ae7ee001f57770a8089b922a4680903918b1/pkg/util/cache/cache.go#L462)):*
> ```Go
> }
> 
> // DoRangeEntryReverse invokes f on all cache entries in the range (to, from]. from
> ```

Shouldn't this be `DoRangeReverseEntry`? They both sound bizarre, but at least that one follows the existing pattern from `llrb.Tree`.


<!-- Sent from Reviewable.io -->

andreimatei committed:
kvclient: improve a test

A test wanted to assert the RangeCache state, but was using the Lookup()
method which, in case of cache miss, goes to the database.

Release note: None

andreimatei reviewed (commented):
> So all things considered, I think my vote would be to store descriptors by EndKey, instead of RangeMetaKey(EndKey) (current behavior) or StartKey (this behavior). I'm curious about what you think.


I think I'll keep the `StartKey`. It's one less thing to remember and I think it's worth it.
I've considered adding support for all the inclusive/exclusive options we need to `OrderedCache`, so that they're hidden at a lower level. I've been lazy so far, but I could...

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52817#-:-MEzGEp69kdeu5Q_6FvY:b-fxxlx0)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/kv/kvclient/kvcoord/range_cache.go, line 558 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/52817#-MEx_B96-bmsSbAMxkB9:-MEy_0F20DIwFsggQpI2:b5q5fnh) ([raw file](https://github.com/cockroachdb/cockroach/blob/a825ae7ee001f57770a8089b922a4680903918b1/pkg/kv/kvclient/kvcoord/range_cache.go#L558)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Instead of scanning backwards and then forwards, can we scan backwards from the EndKey and retain the single pass?
</blockquote></details>

done; see now
___
*[pkg/kv/kvclient/kvcoord/range_cache.go, line 908 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/52817#-MEyJfd398pohUaBgJaY:-MEzG7QJ8D0wB10nY18M:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a825ae7ee001f57770a8089b922a4680903918b1/pkg/kv/kvclient/kvcoord/range_cache.go#L908)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> ```
> rangeCacheKey(roachpb.RKeyMin)
> ```

Should we do the same thing we did with `maxCacheKey`?
</blockquote></details>

done
___
*[pkg/util/cache/cache.go, line 462 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/52817#-MEyIEWoBKyHlLQeElDG:-MEz90cS6z03TFH2MnPq:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a825ae7ee001f57770a8089b922a4680903918b1/pkg/util/cache/cache.go#L462)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Shouldn't this be `DoRangeReverseEntry`? They both sound bizarre, but at least that one follows the existing pattern from `llrb.Tree`.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
Still <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 6 files at r6, 4 of 4 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52817#-:-MF2VYGk8FzoJwlezO0P:b-2pygyf)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52817#-:-MFC_z93DPOZ7TGcwj7a:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->
